StarRailBattleSim/src/characters/AbstractCharacter.java

```diff
@@ -95,41 +104,116 @@ public AbstractCharacter(String name, int baseHP, int baseAtk, int baseDef, int
         this.lightcone = new DefaultLightcone(this);
     }
 
-    public void useSkill() {
+    public void registerGoal(int priority, UltGoal<C> ultGoal) {
+        if (ultGoals.containsKey(priority)) {
+            throw new IllegalArgumentException("Priority already exists, " + priority);
+        }
+        ultGoals.put(priority, ultGoal);
+    }
+
+    public void registerGoal(int priority, TurnGoal<C> turnGoal) {
+        if (turnGoals.containsKey(priority)) {
+            throw new IllegalArgumentException("Priority already exists, " + priority);
+        }
+        turnGoals.put(priority, turnGoal);
+    }
+
+    public final void tryUltimate() {
+        if (currentEnergy < ultCost) {
+            return;
+        }
+
+        for (UltGoal<C> ultGoal : this.ultGoals.values()) {
+            UltGoal.UltGoalResult result = ultGoal.determineAction();
+            switch (result) {
+                case DO: {
+                    getBattle().addToLog(new UltDecision(this, ultGoal.getClass()));
+                    this.ultimateSequence();
+                    return;
+                }
+                case DONT: {
+                    return;
+                }
+                case PASS: {
+                    continue;
+                }
+                default: {
+                    throw new IllegalStateException("Unexpected value: " + result);
+                }
+            }
+        }
+    }
+
+    @Override
+    public final void takeTurn() {
+        super.takeTurn();
+        numTurnsMetric++;
+
+        //tryUltimate();
```

Comment:
I forgot to delete, adding it there increases dmg for most teams (but not all, would need to look at log to write the correct goals so fix that). Was something to discuss, I commented it out to ensure values stay the same in this PR

Reply:
ah right, this would've been the tryUlt at start of turn event. In that case you can leave it commented it out for now, maybe also add a TODO to look into why this event would change the performance of many teams.